### ISSUE TITLE
could not exec command

### ISSUE THREAD

**Issue by xcolwell** [iBotPeaches/Apktool#1083]
On OS X, following the steps at http://ibotpeaches.github.io/Apktool/build/ then running `./gradlew build fatJar` provides the test error below.

Any idea what is wrong here?

```
brut.androlib.ProviderAttributeTest > isProviderStringReplacementWorking FAILED
    brut.androlib.AndrolibException: brut.androlib.AndrolibException: brut.common.BrutException: could not exec command: [/var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/brut_util_Jar_6284933798507256458.tmp, p, --forced-package-id, 127, --min-sdk-version, 21, --target-sdk-version, 22, --version-code, 1, --version-name, 1.0, -F, /var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/APKTOOL7418696414155817342.tmp, -0, arsc, -0, arsc, -I, /Users/brien/Library/apktool/framework/1.apk, -S, /var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/BRUT2407258421040275948.tmp/issue636.apk.out/res, -M, /var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/BRUT2407258421040275948.tmp/issue636.apk.out/AndroidManifest.xml]
        at brut.androlib.Androlib.buildResourcesFull(Androlib.java:472)
        at brut.androlib.Androlib.buildResources(Androlib.java:410)
        at brut.androlib.Androlib.build(Androlib.java:298)
        at brut.androlib.ProviderAttributeTest.isProviderStringReplacementWorking(ProviderAttributeTest.java:59)

        Caused by:
        brut.androlib.AndrolibException: brut.common.BrutException: could not exec command: [/var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/brut_util_Jar_6284933798507256458.tmp, p, --forced-package-id, 127, --min-sdk-version, 21, --target-sdk-version, 22, --version-code, 1, --version-name, 1.0, -F, /var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/APKTOOL7418696414155817342.tmp, -0, arsc, -0, arsc, -I, /Users/brien/Library/apktool/framework/1.apk, -S, /var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/BRUT2407258421040275948.tmp/issue636.apk.out/res, -M, /var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/BRUT2407258421040275948.tmp/issue636.apk.out/AndroidManifest.xml]
            at brut.androlib.res.AndrolibResources.aaptPackage(AndrolibResources.java:425)
            at brut.androlib.Androlib.buildResourcesFull(Androlib.java:458)
            ... 3 more

            Caused by:
            brut.common.BrutException: could not exec command: [/var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/brut_util_Jar_6284933798507256458.tmp, p, --forced-package-id, 127, --min-sdk-version, 21, --target-sdk-version, 22, --version-code, 1, --version-name, 1.0, -F, /var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/APKTOOL7418696414155817342.tmp, -0, arsc, -0, arsc, -I, /Users/brien/Library/apktool/framework/1.apk, -S, /var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/BRUT2407258421040275948.tmp/issue636.apk.out/res, -M, /var/folders/fd/5d3dl8ms00n92tbv_3h5pg0w0000gn/T/BRUT2407258421040275948.tmp/issue636.apk.out/AndroidManifest.xml]
                at brut.util.OS.exec(OS.java:90)
                at brut.androlib.res.AndrolibResources.aaptPackage(AndrolibResources.java:419)
                ... 4 more

58 tests completed, 1 failed
:brut.apktool:apktool-lib:test FAILED
```

**Comment by iBotPeaches** [comment]
Try deleting `$HOME/apktool/framework/1.apk` and rebuilding. Might be outdated framework.

**Comment by xcolwell** [comment]
Thanks - that fixed the usage of apktool 2.0.2. However, I'm still seeing some bugs in 2.0.2 where `apktool b` does not produce a valid APK. When installing the APK produced by `apktool b` (after signing it), I see this error in the logcat: 

```
W/PackageManager( 1090): Failed parse during installPackageLI
W/PackageManager( 1090): android.content.pm.PackageParser$PackageParserException: /data/app/vmdl661512646.tmp/base.apk (at Binary XML file line #152): <meta-data> requires an android:name attribute
W/PackageManager( 1090): 	at android.content.pm.PackageParser.parseBaseApk(PackageParser.java:885)
W/PackageManager( 1090): 	at android.content.pm.PackageParser.parseClusterPackage(PackageParser.java:790)
W/PackageManager( 1090): 	at android.content.pm.PackageParser.parsePackage(PackageParser.java:754)
W/PackageManager( 1090): 	at com.android.server.pm.PackageManagerService.installPackageLI(PackageManagerService.java:10629)
W/PackageManager( 1090): 	at com.android.server.pm.PackageManagerService.access$2300(PackageManagerService.java:238)
W/PackageManager( 1090): 	at com.android.server.pm.PackageManagerService$6.run(PackageManagerService.java:8703)
W/PackageManager( 1090): 	at android.os.Handler.handleCallback(Handler.java:739)
W/PackageManager( 1090): 	at android.os.Handler.dispatchMessage(Handler.java:95)
W/PackageManager( 1090): 	at android.os.Looper.loop(Looper.java:135)
W/PackageManager( 1090): 	at android.os.HandlerThread.run(HandlerThread.java:61)
W/PackageManager( 1090): 	at com.android.server.ServiceThread.run(ServiceThread.java:46)
```


I thought this might be fixed in the master according to the release notes at http://ibotpeaches.github.io/Apktool/changes/. However, when I pull master and try to build following http://ibotpeaches.github.io/Apktool/build/, I see these build errors:

```Executing task ':brut.apktool:apktool-lib:compileJava' (up-to-date check took 0.009 secs) due to:
  No history is available.
All input files are considered out-of-date for incremental task ':brut.apktool:apktool-lib:compileJava'.
Compiling with JDK Java compiler API.
warning: [options] bootstrap class path not set in conjunction with -source 1.7
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/ApktoolProperties.java:19: error: package org.jf.baksmali does not exist
import org.jf.baksmali.baksmali;
                      ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/ApktoolProperties.java:20: error: package org.jf.smali does not exist
import org.jf.smali.main;
                   ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/IndentingWriter.java:24: error: package org.jf.util does not exist
public class IndentingWriter extends org.jf.util.IndentingWriter {
                                                ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:20: error: package org.antlr.runtime does not exist
import org.antlr.runtime.*;
^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:21: error: package org.antlr.runtime.tree does not exist
import org.antlr.runtime.tree.CommonTree;
                             ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:22: error: package org.antlr.runtime.tree does not exist
import org.antlr.runtime.tree.CommonTreeNodeStream;
                             ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:24: error: package org.jf.dexlib2.writer.builder does not exist
import org.jf.dexlib2.writer.builder.DexBuilder;
                                    ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:25: error: package org.jf.smali does not exist
import org.jf.smali.*;
^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:32: error: cannot find symbol
    public static boolean assembleSmaliFile(String smali, DexBuilder dexBuilder, boolean verboseErrors,
                                                          ^
  symbol:   class DexBuilder
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:33: error: cannot find symbol
                                            boolean printTokens, File smaliFile) throws IOException, RuntimeException, RecognitionException {
                                                                                                                       ^
  symbol:   class RecognitionException
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:39: error: cannot find symbol
    public static boolean assembleSmaliFile(InputStream is,DexBuilder dexBuilder, boolean verboseErrors,
                                                           ^
  symbol:   class DexBuilder
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:40: error: cannot find symbol
                                            boolean printTokens, File smaliFile) throws IOException, RecognitionException {
                                                                                                     ^
  symbol:   class RecognitionException
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:53: error: cannot find symbol
    public static boolean assembleSmaliFile(File smaliFile,DexBuilder dexBuilder, boolean verboseErrors,
                                                           ^
  symbol:   class DexBuilder
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:54: error: cannot find symbol
                                            boolean printTokens) throws IOException, RecognitionException {
                                                                                     ^
  symbol:   class RecognitionException
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliBuilder.java:29: error: package org.antlr.runtime does not exist
import org.antlr.runtime.RecognitionException;
                        ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliBuilder.java:31: error: package org.jf.dexlib2.writer.builder does not exist
import org.jf.dexlib2.writer.builder.DexBuilder;
                                    ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliBuilder.java:32: error: package org.jf.dexlib2.writer.io does not exist
import org.jf.dexlib2.writer.io.FileDataStore;
                               ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliBuilder.java:63: error: cannot find symbol
    private void buildFile(String fileName, DexBuilder dexBuilder)
                                            ^
  symbol:   class DexBuilder
  location: class SmaliBuilder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:20: error: package org.jf.baksmali does not exist
import org.jf.baksmali.baksmali;
                      ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:21: error: package org.jf.baksmali does not exist
import org.jf.baksmali.baksmaliOptions;
                      ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:22: error: package org.jf.dexlib2 does not exist
import org.jf.dexlib2.DexFileFactory;
                     ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:23: error: package org.jf.dexlib2.analysis does not exist
import org.jf.dexlib2.analysis.ClassPath;
                              ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:24: error: package org.jf.dexlib2.dexbacked does not exist
import org.jf.dexlib2.dexbacked.DexBackedDexFile;
                               ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:25: error: package org.jf.dexlib2.dexbacked does not exist
import org.jf.dexlib2.dexbacked.DexBackedOdexFile;
                               ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:26: error: package org.jf.dexlib2.analysis does not exist
import org.jf.dexlib2.analysis.InlineMethodResolver;
                              ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/ApktoolProperties.java:54: error: cannot find symbol
            templateStream = baksmali.class.getClassLoader().getResourceAsStream("baksmali.properties");
                             ^
  symbol:   class baksmali
  location: class ApktoolProperties
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/ApktoolProperties.java:72: error: cannot find symbol
            templateStream = main.class.getClassLoader().getResourceAsStream("smali.properties");
                             ^
  symbol:   class main
  location: class ApktoolProperties
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:56: error: cannot find symbol
        CommonTokenStream tokens;
        ^
  symbol:   class CommonTokenStream
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:57: error: cannot find symbol
        LexerErrorInterface lexer;
        ^
  symbol:   class LexerErrorInterface
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:62: error: cannot find symbol
        lexer = new smaliFlexLexer(reader);
                    ^
  symbol:   class smaliFlexLexer
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:63: error: cannot find symbol
        ((smaliFlexLexer)lexer).setSourceFile(smaliFile);
          ^
  symbol:   class smaliFlexLexer
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:64: error: cannot find symbol
        tokens = new CommonTokenStream((TokenSource) lexer);
                     ^
  symbol:   class CommonTokenStream
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:64: error: cannot find symbol
        tokens = new CommonTokenStream((TokenSource) lexer);
                                        ^
  symbol:   class TokenSource
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:70: error: cannot find symbol
                Token token = tokens.get(i);
                ^
  symbol:   class Token
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:71: error: cannot find symbol
                if (token.getChannel() == smaliParser.HIDDEN) {
                                          ^
  symbol:   variable smaliParser
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:75: error: cannot find symbol
                System.out.println(smaliParser.tokenNames[token.getType()] + ": " + token.getText());
                                   ^
  symbol:   variable smaliParser
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:79: error: cannot find symbol
        smaliParser parser = new smaliParser(tokens);
        ^
  symbol:   class smaliParser
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:79: error: cannot find symbol
        smaliParser parser = new smaliParser(tokens);
                                 ^
  symbol:   class smaliParser
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:82: error: package smaliParser does not exist
        smaliParser.smali_file_return result = parser.smali_file();
                   ^
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:88: error: cannot find symbol
        CommonTree t = (CommonTree) result.getTree();
        ^
  symbol:   class CommonTree
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:88: error: cannot find symbol
        CommonTree t = (CommonTree) result.getTree();
                        ^
  symbol:   class CommonTree
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:90: error: cannot find symbol
        CommonTreeNodeStream treeStream = new CommonTreeNodeStream(t);
        ^
  symbol:   class CommonTreeNodeStream
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:90: error: cannot find symbol
        CommonTreeNodeStream treeStream = new CommonTreeNodeStream(t);
                                              ^
  symbol:   class CommonTreeNodeStream
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:93: error: cannot find symbol
        smaliTreeWalker dexGen = new smaliTreeWalker(treeStream);
        ^
  symbol:   class smaliTreeWalker
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/mod/SmaliMod.java:93: error: cannot find symbol
        smaliTreeWalker dexGen = new smaliTreeWalker(treeStream);
                                     ^
  symbol:   class smaliTreeWalker
  location: class SmaliMod
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliBuilder.java:52: error: cannot find symbol
            DexBuilder dexBuilder = DexBuilder.makeDexBuilder();
            ^
  symbol:   class DexBuilder
  location: class SmaliBuilder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliBuilder.java:52: error: cannot find symbol
            DexBuilder dexBuilder = DexBuilder.makeDexBuilder();
                                    ^
  symbol:   variable DexBuilder
  location: class SmaliBuilder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliBuilder.java:57: error: cannot find symbol
            dexBuilder.writeTo(new FileDataStore( new File(mDexFile.getAbsolutePath())));
                                   ^
  symbol:   class FileDataStore
  location: class SmaliBuilder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliBuilder.java:73: error: cannot find symbol
            } catch (IOException | RecognitionException ex) {
                                   ^
  symbol:   class RecognitionException
  location: class SmaliBuilder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliBuilder.java:115: error: cannot find symbol
        } catch (IOException | RecognitionException ex) {
                               ^
  symbol:   class RecognitionException
  location: class SmaliBuilder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:62: error: cannot find symbol
            ClassPath.dontLoadClassPath = mDebug;
            ^
  symbol:   variable ClassPath
  location: class SmaliDecoder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:64: error: cannot find symbol
            baksmaliOptions options = new baksmaliOptions();
            ^
  symbol:   class baksmaliOptions
  location: class SmaliDecoder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:64: error: cannot find symbol
            baksmaliOptions options = new baksmaliOptions();
                                          ^
  symbol:   class baksmaliOptions
  location: class SmaliDecoder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:76: error: cannot find symbol
            options.registerInfo = (mDebug ? baksmaliOptions.DIFFPRE : 0);
                                             ^
  symbol:   variable baksmaliOptions
  location: class SmaliDecoder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:94: error: cannot find symbol
            DexBackedDexFile dexFile = DexFileFactory.loadDexFile(mApkFile, mDexFile, mApi, false);
            ^
  symbol:   class DexBackedDexFile
  location: class SmaliDecoder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:94: error: cannot find symbol
            DexBackedDexFile dexFile = DexFileFactory.loadDexFile(mApkFile, mDexFile, mApi, false);
                                       ^
  symbol:   variable DexFileFactory
  location: class SmaliDecoder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:100: error: cannot find symbol
            if (dexFile instanceof DexBackedOdexFile) {
                                   ^
  symbol:   class DexBackedOdexFile
  location: class SmaliDecoder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:102: error: cannot find symbol
                        InlineMethodResolver.createInlineMethodResolver(((DexBackedOdexFile)dexFile).getOdexVersion());
                                                                          ^
  symbol:   class DexBackedOdexFile
  location: class SmaliDecoder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:102: error: cannot find symbol
                        InlineMethodResolver.createInlineMethodResolver(((DexBackedOdexFile)dexFile).getOdexVersion());
                        ^
  symbol:   variable InlineMethodResolver
  location: class SmaliDecoder
/Users/brien/Apktool/brut.apktool/apktool-lib/src/main/java/brut/androlib/src/SmaliDecoder.java:105: error: cannot find symbol
            baksmali.disassembleDexFile(dexFile,options);
            ^
  symbol:   variable baksmali
  location: class SmaliDecoder
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
60 errors
1 warning
:brut.apktool:apktool-lib:compileJava FAILED
```

**Comment by iBotPeaches** [comment]
That looks like `smali` patches weren't applied. Ensure that the `brut.apktool.smali` directory has contents.

If not.

1. Run `git submodule update --init --recursive` to bring in the submodule
2. Run `./gradlew applyPatches` to create `brut.apktool.smali`

Since the original issue and this secondary issue are not bugs, but build configuration troubles. I am closing this.
